### PR TITLE
BeneficiariesFile now expects a rational number, which can be scaled …

### LINKED ISSUE #18: Add a `decimal places` flag, that multiples all values by 10^places
Since we give constraints in terms of integer underlying tokens, and clients usually give in terms of float multiplied tokens (so 1 of token X where it has 6dp, is actually 1000000 of integer token X), we should add a way to handle this automatically.

### PR DESCRIPTION
**Pull request by jy14898**
…by the --decimal-places flag

Closes https://github.com/mlabs-haskell/CardStarter-bulk-airdrop-script/issues/18.

Should we warn on truncation?

### CODE REVIEW

**Comment by samuelWilliams99** [comment]
Truncation is an annoying one - i think we need to be REALLY explicit here.
This will extend the task a little bit, but I think the safest bet is to completely fail parsing on truncation - UNLESS, we have a flag that allows it.
This flag should be very explicit that is always rounds down.
So, to summarise:
- Add a truncate flag - make it obvious that it rounds down (using the help text)
- If we try to truncate a value and this flag is not set, fail parsing.

File: src/BeneficiariesFile.hs [hunk 1]
```diff
@@ -50,16 +51,19 @@ parseBeneficiary conf = toBeneficiary . words
       makeBeneficiary addr (maybeToMissing "quantity" conf.dropAmount) (maybeToMissing "assetclass" conf.assetClass)
     toBeneficiary _ = Left "Invalid number of inputs"
 
-    makeBeneficiary :: Text -> Either Text Integer -> Either Text AssetClass -> Either Text Beneficiary
-    makeBeneficiary addr eAmt eAc = Beneficiary <$> parseAddress conf.usePubKeys addr <*> eAmt <*> eAc
+    scaleAmount :: Ratio Integer -> Integer
+    scaleAmount r = numerator r * conf.decimalPlaces `div` denominator r
```
Comment (samuelWilliams99) [inline src/BeneficiariesFile.hs:55]: does multiplying the numerator by the decimal places do what we want? should it not be something like `10^decimal places`?

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]

**Comment by jy14898** [comment]
I've moved from Rational to Scientific for representing pre-integer token values because:
 - Scientific more naturally represents what we're doing here (c * (10^e))
 - Actually parses correctly

I've added the `--truncate` flag to allow dropping change when rounding down. I've also added a test which checks if the truncation occurs only when it should (regardless of the flag setting).

**Review by samuelWilliams99 — APPROVED** [review]

File: src/BeneficiariesFile.hs [hunk 2]
```diff
@@ -50,16 +51,32 @@ parseBeneficiary conf = toBeneficiary . words
...
+    makeBeneficiary addr eAmt eAc = Beneficiary <$> parseAddress conf.usePubKeys addr <*> (eAmt >>= scaleAmount) <*> eAc
+
+-- Converts a Scientific to an Integer. If truncation occurs, return the truncated value and the change
+scientificToInteger :: Scientific -> (Integer, Maybe Scientific)
+scientificToInteger s = either (const truncated) (,Nothing) (floatingOrInteger @Float s)
+  where
+    b10e = base10Exponent s
+
+    -- For truncated to be evaluated, the exponent must have been negative
+    truncated = Just . (*) (10 ^^ b10e) . fromInteger <$> coefficient s `divMod` (10 ^ negate b10e)
```
[the remark this answers is not in the document]
Comment (jy14898) [reply src/BeneficiariesFile.hs]: I've just pushed another go at it, no longer using point free/tuple functor instance

[the remark this answers is not in the document]
Comment (samuelWilliams99) [reply src/BeneficiariesFile.hs]: Thats looks better, thanks! Happy to merge when this passes CI

**Pull request merged**